### PR TITLE
limchengyang/add_callback_log_appid

### PR DESCRIPTION
**Pull request by limchengyang**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -232,6 +232,12 @@ Global hook which will run immediately before send API response.
 
 Global hook which will run immediately after sent API response back to client.
 
+##### log\_error\_on\_finish (global)
```
Comment (zakame) [inline README.md]: Shouldn't this be `on_error` instead?

File: README.md [hunk 2]
```diff
@@ -232,6 +232,12 @@ Global hook which will run immediately before send API response.
...
+
+    log_error_on_finish => [sub { my ($d, $code, $error_msg) = @_; ... }, sub {...}]
```
Comment (zakame) [inline README.md]: ditto, also it seems it accepts only a single coderef rather than a series of coderefs.

File: README.md (hunk 1 above)
Comment (limchengyang) [reply README.md]: oops mybad, forgot to update this to latest change. Have just updated them accordingly. Thanks for spotting !

File: README.md (hunk 2 above)
Comment (limchengyang) [reply README.md]: same here

File: lib/Mojo/WebSocketProxy/Dispatcher.pm [hunk 3]
```diff
@@ -69,20 +69,41 @@ sub open_connection {
...
+                    request_body    => $msg,
+                },
+                }) if exists $config->{on_error};
             return;
         };
 
         my $normalized_msg = eval { Unicode::Normalize::NFC($decoded) } or do {
-            stats_inc("websocket_proxy.unicode_normalisation.failure", {tags => ['error_code:1007']});
-            $c->finish(1007 => 'Malformed Unicode data');
+            $config->{on_error}({
```
Comment (binarytom) [inline lib/Mojo/WebSocketProxy/Dispatcher.pm]: Can you use `->emit` instead? Mojolicious already has some basic event handling built in.

See https://github.com/binary-com/perl-Mojo-WebSocketProxy/blob/master/lib/Mojo/WebSocketProxy/Parser.pm#L52 for an example.

File: Changes [hunk 4]
```diff
@@ -2,6 +2,12 @@ Revision history for Perl module Mojo::WebSocketProxy
 
 {{$NEXT}}
 
+0.12      2019-05-06 16:57:54+08:00 Asia/Minila
```
Comment (zakame) [inline Changes]: With this line and version generated, has the resulting dist been uploaded to PAUSE already?

Comment (zakame) [reply Changes]: Also I don't see any corresponding `Makefile.PL` changes

Comment (limchengyang) [reply Changes]: our bad, first time working on public module, didnt know that we should upgrade the versioning in another PR with `dzil release`. We somehow did this manually, removing it right now 😅

File: lib/Mojo/WebSocketProxy/Dispatcher.pm (hunk 3 above)
Comment (limchengyang) [reply lib/Mojo/WebSocketProxy/Dispatcher.pm]: changes are made based on the comment

File: lib/Mojo/WebSocketProxy/Dispatcher.pm [hunk 5]
```diff
@@ -68,21 +68,19 @@ sub open_connection {
...
         # Incoming data will be JSON-formatted text, as a Unicode string.
         # We normalize the entire string before decoding.
-        my $decoded = eval { Encode::decode_utf8($msg) } or do {
-            stats_inc("websocket_proxy.utf8_decoding.failure", {tags => ['error_code:1007']});
-            $c->finish(1007 => 'Malformed UTF-8 data');
+
+        my $decoded = eval { Encode::decode_utf8($msg, Encode::FB_CROAK) } or do {
+            $c->tx->emit(encoding_error => _get_error_details(code => 'INVALID_UTF8', reason => 'Malformed UTF-8 data', message => $msg));
             return;
         };
```
Comment (limchengyang) [inline lib/Mojo/WebSocketProxy/Dispatcher.pm]:
```suggestion
        # The Unicode check is added as a safety net. However, the error is not triggered so far.
```

File: lib/Mojo/WebSocketProxy/Dispatcher.pm [hunk 6]
```diff
@@ -68,21 +68,19 @@ sub open_connection {
...
-
+        # The Unicode check is added as a safety net. However, the error is not triggered so far.
```
Comment (limchengyang) [inline lib/Mojo/WebSocketProxy/Dispatcher.pm]:
```suggestion

        # The Unicode::Normalize::NFC check is added as a safety net. However, the error is not triggered so far.
```

**Pull request merged**